agent/stats/task.go

```diff
@@ -0,0 +1,155 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/stats/resolver"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+
+	"github.com/cihub/seelog"
+	"github.com/docker/docker/api/types"
+	dockerstats "github.com/docker/docker/api/types"
+)
+
+// statsTaskCommon contains the common fields in StatsTask for both Linux and Windows.
+// StatsTask abstracts methods to gather and aggregate network data for a task. Used only for AWSVPC mode.
+type statsTaskCommon struct {
+	StatsQueue            *Queue
+	TaskMetadata          *TaskMetadata
+	Ctx                   context.Context
```

Comment:
This really is a golang anit-pattern to expose the ctx and cancel. Typically the caller would provide these and you can consume them here. So if we wanted to do this there would be a:

`func NewStatsTask(ctx context)` ... and then we would just wait entirely on this context to be done. The caller is then in control of when to exit etc based on its logic.

Reply:
So, `StatsTask` is responsible for the entire lifecycle of stats collection of the container associated with one instance of StatsTask. When the stats collection needs to be stopped due to any reason, the `StopStatsCollection` method would be invoked on the object which in turn would cancel the context.

Reference- https://github.com/aws/amazon-ecs-agent/blob/fe79b107f3b204ca1a831b1d15424d12d2948bdf/agent/stats/task_linux.go#L116

Reply:
I know you didnt change this code so just leave it for now thats fine.